aptos-move/framework/supra-framework/sources/automation_registry_state.move

```diff
@@ -0,0 +1,623 @@
+        if (automation_task_metadata.state == PENDING) {
+            enumerable_map::remove_value(&mut state.tasks, id);
+        } else if (automation_task_metadata.state == ACTIVE) {
+            automation_task_metadata.state = CANCELLED;
+            enumerable_map::update_value(&mut state.tasks, id, automation_task_metadata);
+        };
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the cancelled task
+        state.gas_committed_for_next_epoch = state.gas_committed_for_next_epoch - automation_task_metadata.max_gas_amount;
+
+        event::emit(CanclledAutomationTask { id: automation_task_metadata.id });
+        automation_task_metadata
+    }
+
+    /// Update Automation gas limit
+    public (friend) fun update_automation_gas_limit(
+        supra_framework: &signer,
+        automation_gas_limit: u64
+    ) acquires AutomationRegistryState {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let state = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        assert!(state.gas_committed_for_next_epoch < automation_gas_limit, EUNACCEPTABLE_AUTOMATION_GAS_LIMIT);
+
+        state.automation_gas_limit = automation_gas_limit;
+
+        event::emit(UpdateAutomationGasLimit { automation_gas_limit });
+    }
+
+    /// List all the automation task ids
+    public(friend) fun get_active_task_ids(): vector<u64> acquires AutomationRegistryState {
+        let state = borrow_global<AutomationRegistryState>(@supra_framework);
+
+        let active_task_ids = vector[];
+        let ids = enumerable_map::get_map_list(&state.tasks);
+
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value(&state.tasks, id);
+            if (task.state != PENDING) {
+                vector::push_back(&mut active_task_ids, id);
+            };
+        });
+        return active_task_ids
+    }
+
+    /// Retrieves the details of a automation task entry by its ID.
+    /// Error will be returned if entry with specified ID does not exist.
+    public (friend) fun get_task_details(id: u64): AutomationTaskMetaData acquires AutomationRegistryState {
+        let automation_task_metadata = borrow_global<AutomationRegistryState>(@supra_framework);
+        assert!(enumerable_map::contains(&automation_task_metadata.tasks, id), EAUTOMATION_TASK_NOT_FOUND);
+        enumerable_map::get_value(&automation_task_metadata.tasks, id)
+    }
+
+    /// Checks whether there is an active task in registry with specified input task id.
+    public(friend) fun has_active_task_with_id(id: u64): bool acquires AutomationRegistryState {
+        let automation_task_metadata = borrow_global<AutomationRegistryState>(@supra_framework);
+        if (enumerable_map::contains(&automation_task_metadata.tasks, id)) {
+            let value = enumerable_map::get_value(&automation_task_metadata.tasks, id);
+            value.state != PENDING
```

Comment:
```suggestion
            value.state == ACTIVE
```
There is a `CANCELLED` state as well which would be missed by the existing logic

Reply:
As agreed, the task cancelled during epoch N is still active during that epoch, and only removed upon new epoch start.
so if node restarts in the middle of the epoch it should be able to get correct list of the tasks that are active during that epoch. that's why here we have `!= PENDING`

Reply:
Fair enough